src/storage/src/upsert_continual_feedback.rs

```diff
@@ -0,0 +1,826 @@
+                    if !hydrating {
+                        tracing::debug!(persist_stash = %persist_stash.len(), ?persist_upper, "ingesting state updates from persist");
+
+                        ingest_state_updates::<_, G, _, _, _>(
+                            &mut persist_stash,
+                            &persist_upper,
+                            &mut error_emitter,
+                            &mut state,
+                            ).await;
+                    }
+                }
+                input_event = input.next() => {
+                    let Some(input_event) = input_event else {
+                        tracing::debug!("input exhausted, shutting down");
+                        break;
+                    };
+                    // Buffer as many events as possible. This should be bounded, as new data can't be
+                    // produced in this worker until we yield to timely.
+                    let events = [input_event]
+                        .into_iter()
+                        .chain(std::iter::from_fn(|| input.next().now_or_never().flatten()))
+                        .enumerate();
+
+                    for (i, event) in events {
+                        tracing::trace!(?event, "source input");
+
+                        match event {
+                            AsyncEvent::Data(cap, mut data) => {
+                                tracing::trace!(
+                                    time=?cap.time(),
+                                    updates=%data.len(),
+                                    "received data in upsert"
+                                );
+
+                                let event_time = cap.time().clone();
+
+                                stage_input(
+                                    &mut stash,
+                                    cap,
+                                    &mut data,
+                                    &input_upper,
+                                    &resume_upper,
+                                    upsert_config.shrink_upsert_unused_buffers_by_ratio,
+                                );
+
+                                // For the very first snapshot from the source,
+                                // we allow emitting partial updates without yet
+                                // seeing all updates for that timestamp. For
+                                // this, we use an ephemeral upsert state to run
+                                // our upserting logic that we throw away once
+                                // we receive the first updates on the persist
+                                // input.
+                                //
+                                // We can only do this for the very first
+                                // snapshot, because all subsequent emitted
+                                // updates need to have a consistent view of
+                                // upsert state as ingested from the persist
```

Comment:
It sounds like we need to change our upsert state to store this type for each key:
```rust
enum UpsertValue {
    value: Value,
    provisional_value: Option<Value>,
}
```
so that we can always emit partial updates and perform the upsert logic using the `provisional_value` of our state and once we receive the final settled values from persist we update both to that value and move on to the next timestamp